test/e2e/src/test-swap.js

```diff
@@ -1,183 +1,148 @@
-    "swap",
-    { keypair: root },
-  );
-  await (
-    await tokenA.initialize({
-      admin: root.publicKey(),
-      decimal: 0,
-      name: "Token A",
-      symbol: "A",
-    })
-  ).signAndSend();
-  await (
-    await tokenA.mint({ amount: amountAToSwap, to: alice.publicKey() })
-  ).signAndSend();
-
-  await (
-    await tokenB.initialize({
-      admin: root.publicKey(),
-      decimal: 0,
-      name: "Token B",
-      symbol: "B",
-    })
-  ).signAndSend();
-  await (
-    await tokenB.mint({ amount: amountBToSwap, to: bob.publicKey() })
-  ).signAndSend();
-
-  t.context = {
-    // eslint-disable-line no-param-reassign
-    root,
-    alice,
-    bob,
-    swapContractAsRoot,
-    swapId,
-    tokenA,
-    tokenAId,
-    tokenB,
-    tokenBId,
-  };
-});
+describe("Swap Contract Tests", function () {
 
-test("calling `signAndSend()` too soon throws descriptive error", async (t) => {
-  const tx = await t.context.swapContractAsRoot.swap({
-    a: t.context.alice.publicKey(),
-    b: t.context.bob.publicKey(),
-    token_a: t.context.tokenAId,
-    token_b: t.context.tokenBId,
-    amount_a: amountAToSwap,
-    min_a_for_b: amountAToSwap,
-    amount_b: amountBToSwap,
-    min_b_for_a: amountBToSwap,
-  });
-  const error = await t.throwsAsync(tx.signAndSend());
-  t.true(
-    error instanceof contract.AssembledTransaction.Errors.NeedsMoreSignatures,
-    `error is not of type 'NeedsMoreSignaturesError'; instead it is of type '${error?.constructor.name}'`,
-  );
-  if (error) t.regex(error.message, /needsNonInvokerSigningBy/);
-});
+  before(async function () {
+    const alice = await generateFundedKeypair();
+    const bob = await generateFundedKeypair();
 
-test("alice swaps bob 10 A for 1 B", async (t) => {
-  const tx = await t.context.swapContractAsRoot.swap({
-    a: t.context.alice.publicKey(),
-    b: t.context.bob.publicKey(),
-    token_a: t.context.tokenAId,
-    token_b: t.context.tokenBId,
-    amount_a: amountAToSwap,
-    min_a_for_b: amountAToSwap,
-    amount_b: amountBToSwap,
-    min_b_for_a: amountBToSwap,
-  });
+    const {
+      client: tokenA,
+      contractId: tokenAId,
+      keypair: root,
+    } = await clientFor("token");
+    const { client: tokenB, contractId: tokenBId } = await clientFor("token", {
+      keypair: root,
+    });
+    const { client: swapContractAsRoot, contractId: swapId } = await clientFor(
+      "swap",
+      { keypair: root },
+    );
+    await (
+      await tokenA.initialize({
+        admin: root.publicKey(),
+        decimal: 0,
+        name: "Token A",
+        symbol: "A",
+      })
+    ).signAndSend();
+    await (
+      await tokenA.mint({ amount: amountAToSwap, to: alice.publicKey() })
+    ).signAndSend();
+
+    await (
+      await tokenB.initialize({
+        admin: root.publicKey(),
+        decimal: 0,
+        name: "Token B",
+        symbol: "B",
+      })
+    ).signAndSend();
```

Comment:
gross :cry: would `.then()`s be cleaner? I'm not suggesting a refactor, just starting a discussion for best practices we can recommend

Reply:
Yeah, `.then` would probably look cleaner 

```suggestion
    await tokenB.initialize({
      admin: root.publicKey(),
      decimal: 0,
      name: "Token B",
      symbol: "B",
    }).then(t => t.signAndSend());
```